packages/process/src/node/raw-process.ts

```diff
@@ -60,6 +62,8 @@ export class RawProcess extends Process {
     kill(signal?: string) {
         if (this.killed === false) {
             this.process.kill(signal);
+            this.processManager.unregister(this);
```

Comment:
This is wrong here for a few reasons:
   - kill may have another signal than SIGTERM here, the user may send SIGHUP for example and thus the process shoud not be unregistred.

  - We should not unregister the process on SIGTERM to keep history, and if we did for now do that we would need to leave the kill to the unregister call, othewise we call kill twice and that could mean sending a kill signal to the wrong pid.

 - I think the LSP contribution should clean up it's processes by calling processManager.unregister(process)

 - That should be the general way to unregister a process. Otherwise we could add a Process->unregister() function that calls the processManager->unregister()

Reply:
> We should not unregister the process on SIGTERM to keep history

That is a very good idea, but this PR has no intention doing anything with the history right now.

> I think the LSP contribution should clean up it's processes by calling processManager.unregister(process)

I think it would be odd. So when I create a process via the factory, then the process is registered into the system. Registration is done "automagically" but the unregister code should be invoked explicitly.

Reply:
Yes I also this it's odd.

I think the best solution would be as described a bit in : https://github.com/theia-ide/theia/pull/841#discussion_r152328076

- In Process class register a on('exit') callback  and call processManager->unregister()
- Kill doesn't call unregister

That way processes will be cleaned-up automatically and terminal references will get notified of that using the delete event from processManager

It will also avoid checking for the signal type etc in kill.

Reply:
Oops sorry it needs to be the ProcessManager that registers the on('exit') callback otherwise the Process will have a reference to itself and not get cleaned up.

Reply:
> In Process class register a on('exit') callback and call processManager->unregister()

What do you think, shall we remove the on('exit') => unregister from the `TerminalBackendContribution`? Seems to be a duplicate.

Reply:
I don't see such a callback ? You mean ws.on('close' ? this one is used to kill the process.

We could change it to kill once the auto unregister is done.